File: skillful_network_server/src/main/java/fr/uca/cdr/skillful_network/model/entities/Simulation.java
package fr.uca.cdr.skillful_network.model.entities;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import fr.uca.cdr.skillful_network.model.entities.simulation.exercise.Exam;
import fr.uca.cdr.skillful_network.model.entities.simulation.exercise.Result;
import fr.uca.cdr.skillful_network.request.SimulationForm;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.PastOrPresent;
import java.util.Date;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "simulations")
public class Simulation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "user_id")
    @JsonIgnoreProperties("simulationSet")
    private User user;

    @NotNull(message = "Job goal cannot be null")
    private String jobGoal;

    @PastOrPresent
    private Date creationDate;
    
    private String synthesis;
    
    @OneToMany(cascade = CascadeType.ALL)
    @JoinColumn(name = "result_id")
    private Set<Result> results; 
    
    
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "jobOffer_id")
    private JobOffer jobOffer;
    
    private boolean jobAccess;
    
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "training_id")
    private Training training;
    
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "exam_id")
    private Exam exam;

    @Transient
    private SimulationForm simulationForm;
    
    public Simulation() {
    }

    public Simulation(@NotNull(message = "Job goal cannot be null") String jobGoal) {
        this.jobGoal = jobGoal;
        this.creationDate = new Date();
        this.jobAccess = false;
    }

    public Simulation(@NotNull(message = "Job goal cannot be null") String jobGoal, @PastOrPresent Date creationDate, String synthesis) {
        this.jobGoal = jobGoal;
        this.creationDate = creationDate;
        this.synthesis = synthesis;
        this.jobAccess = false;
    }

    public Simulation(long id, @NotNull(message = "Job goal cannot be null") String jobGoal, @PastOrPresent Date creationDate, String synthesis) {
        this.id = id;
        this.jobGoal = jobGoal;
        this.creationDate = creationDate;
        this.synthesis = synthesis;
        this.jobAccess = false;
    }
    
    public Simulation(@NotNull(message = "Job goal cannot be null") String jobGoal, JobOffer jobOffer) {
    	this.jobGoal = jobGoal;
    	this.jobOffer = jobOffer;
        this.creationDate = new Date();
        this.jobAccess = false;
    }
    
    public Simulation(@NotNull(message = "Job goal cannot be null") String jobGoal, JobOffer jobOffer, Exam exam) {
    	this.jobGoal = jobGoal;
    	this.jobOffer = jobOffer;
        this.creationDate = new Date();
        this.jobAccess = false;
        this.exam = exam;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public User getUser() { return user; }

    public void setUser(User user) { this.user = user; }

    public String getJobGoal() {
        return jobGoal;
    }

    public void setJobGoal(String jobGoal) {
        this.jobGoal = jobGoal;
    }

    public Date getCreationDate() {
        return creationDate;
    }

    public void setCreationDate(Date creationDate) {
        this.creationDate = creationDate;
    }

    public String getSynthesis() {
        return synthesis;
    }

    public void setSynthesis(String synthesis) {
        this.synthesis = synthesis;
    }

    public Set<Result> getResults() {
		return results;
	}

	public void setResults(Set<Result> results) {
		this.results = results;
	}

	@Override
    public int hashCode() {
        return Objects.hash(id, jobGoal);
    }
	
    public JobOffer getJobOffer() {
		return jobOffer;
	}

	public void setJobOffer(JobOffer jobOffer) {
		this.jobOffer = jobOffer;
	}

	public boolean isJobAccess() {
		return jobAccess;
	}

	public void setJobAccess(boolean jobAccess) {
		this.jobAccess = jobAccess;
	}

	public Training getTraining() {
		return training;
	}

	public void setTraining(Training training) {
		this.training = training;
	}

	public Exam getExam() {
		return exam;
	}

	public void setExam(Exam exam) {
		this.exam = exam;
	}

	public SimulationForm getSimulationForm() {
		return simulationForm;
	}

	public void setSimulationForm(SimulationForm simulationForm) {
		this.simulationForm = simulationForm;
	}

	@Override
	public String toString() {
		return "Simulation [id=" + id + ", user=" + user + ", jobGoal=" + jobGoal + ", creationDate=" + creationDate
				+ ", synthesis=" + synthesis + ", results=" + results + ", jobOffer=" + jobOffer + ", jobAccess="
				+ jobAccess + ", training=" + training + ", exam=" + exam + ", simulationForm=" + simulationForm + "]";
	}

//	@Override
//    public String toString() {
//        return "Simulation [" + "id=" + id +
//                "] userID=" + user.getId() +
//                ", jobGoal=" + jobGoal +
//                ", creationDate=" + creationDate +
//                ", synthesis=" + synthesis;
//    }
	
}

File: skillful_network_client/src/app/shared/models/simulation.ts
import { User } from './user';
import { JobOffer } from './job-offer';
import { Training } from './training';
import { IExam, MOCK_EXAM, IExamForm, MOCK_EXAM_FORM } from './mock.examen';

export class Simulation {

    private _id: number;
    private _userId: number;
    private _jobGoal: String;
    private _creationDate: Date ;
    private _synthesis: String ;
    private _jobOffer: JobOffer;
    private _jobAccess: boolean;
    private _training: Training;
    private _exam: IExam = MOCK_EXAM;
    private _examForm: IExamForm  = MOCK_EXAM_FORM;

    constructor(data: any) {
        this._id = data.id;
        this._userId = data.user.id;
        this._jobGoal = data.jobGoal;
        this._creationDate = data.creationDate;
        this._synthesis = data.synthesis;
        this._jobOffer = data.jobOffer;
        this._jobAccess = data.jobAccess;
        this._training = data.training;
        this._exam = data.exam;
        this._examForm = data.simulationForm;
    }
    
    public get id(): number {
        return this._id;
    }
    public set id(value: number) {
        this._id = value;
    }
    public get userId(): number {
        return this._userId;
    }
    public set userId(value: number) {
        this._userId = value;
    }
    public get jobGoal(): String {
        return this._jobGoal;
    }
    public set jobGoal(value: String) {
        this._jobGoal = value;
    }
    public get creationDate(): Date {
        return this._creationDate;
    }
    public set creationDate(value: Date) {
        this._creationDate = value;
    }
    public get synthesis(): String {
        return this._synthesis;
    }
    public set synthesis(value: String) {
        this._synthesis = value;
    }
    public get jobOffer(): JobOffer {
        return this._jobOffer;
    }
    public set jobOffer(value: JobOffer) {
        this._jobOffer = value;
    }
    public get training(): Training {
        return this._training;
    }
    public set training(value: Training) {
        this._training = value;
    }
    public get jobAccess(): boolean {
        return this._jobAccess;
    }
    public set jobAccess(value: boolean) {
        this._jobAccess = value;
    }
    public get exam(): IExam {
        return this._exam;
    }
    public set exam(value: IExam) {
        this._exam = value;
    }
    public get examForm(): IExamForm {
        return this._examForm;
    }
    public set examForm(value: IExamForm) {
        this._examForm = value;
    }

}
